endpoints/events/event.go

```diff
@@ -0,0 +1,364 @@
+		0x00, 0x00, 0x00, 0x0D, 0x49, 0x44, 0x41, 0x54, 0x08, 0x99, 0x63, 0x60, 0x60, 0x60, 0x60, 0x00, 0x00,
+		0x00, 0x05, 0x00, 0x01, 0x87, 0xA1, 0x4E, 0xD4, 0x00, 0x00, 0x00, 0x00, 0x49, 0x45, 0x4E, 0x44, 0xAE,
+		0x42, 0x60, 0x82},
+	ContentType: "image/png",
+}
+
+type EventRequest struct {
+	Type      EventType      `json:"type,omitempty"`
+	Format    ResponseFormat `json:"format,omitempty"`
+	Analytics Analytics      `json:"analytics,omitempty"`
+	Bidid     string         `json:"bidid,omitempty"`
+	AccountID string         `json:"account_id,omitempty"`
+	Bidder    string         `json:"bidder,omitempty"`
+	Timestamp int64          `json:"timestamp,omitempty"`
+}
+
+type TrackingPixel struct {
+	Content     []byte `json:"content,omitempty"`
+	ContentType string `json:"content_type,omitempty"`
+}
+
+type eventEndpoint struct {
+	Accounts      stored_requests.AccountFetcher
+	Analytics     analytics.PBSAnalyticsModule
+	Cfg           *config.Configuration
+	TrackingPixel *TrackingPixel
+}
+
+func NewEventEndpoint(cfg *config.Configuration, accounts stored_requests.AccountFetcher, analytics analytics.PBSAnalyticsModule) httprouter.Handle {
+	ee := &eventEndpoint{
+		Accounts:      accounts,
+		Analytics:     analytics,
+		Cfg:           cfg,
+		TrackingPixel: trackingPixelPng,
+	}
+
+	return ee.Handle
+}
+
+func (e *eventEndpoint) Handle(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	// parse event request from http req
+	eventRequest, err := ParseEventRequest(r)
+
+	// handle possible parsing errors
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(fmt.Sprintf("invalid request: %s\n", err.Error())))
+
+		return
+	}
+
+	// validate account id
+	accountId, err := validateRequiredParameter(r, AccountIdParameter)
+
+	if err != nil {
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte(fmt.Sprintf("Account '%s' is required query parameter and can't be empty", AccountIdParameter)))
+		return
+	}
+	eventRequest.AccountID = accountId
+
+	if eventRequest.Analytics != Enabled {
+		return
+	}
+
+	ctx := context.Background()
+
+	// get account details
+	account, errs := GetAccount(ctx, e.Cfg, e.Accounts, eventRequest.AccountID)
+	if len(errs) > 0 {
+		w.WriteHeader(http.StatusInternalServerError)
+		for _, err := range errs {
+			w.Write([]byte(fmt.Sprintf("Internal Error: %s\n", err.Error())))
+		}
+
+		return
+	}
+
+	// account does not support events
+	if !account.EventsEnabled {
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte(fmt.Sprintf("Account '%s' doesn't support events", eventRequest.AccountID)))
+		return
+	}
+
+	// handle notification event
+	e.Analytics.LogNotificationEventObject(&analytics.NotificationEvent{
+		Type:      string(eventRequest.Type),
+		Timestamp: time.Unix(eventRequest.Timestamp, 0),
+		Bidid:     eventRequest.Bidid,
+		Bidder:    eventRequest.Bidder,
+		Account:   account,
```

Comment:
Passing the entire `EventRequest` object instead of select fields would be more consistent with the other analytics methods.

Reply:
The implementation follows the pattern used on both PBS Java and PBS Go: 

1. PBS Java
1.1. https://github.com/rubicon-project/prebid-server-java/blob/master/src/main/java/org/prebid/server/handler/NotificationEventHandler.java#L112

2. PBS Go
2.1. https://github.com/prebid/prebid-server/blob/master/endpoints/openrtb2/amp_auction.go#L90
2.2. https://github.com/prebid/prebid-server/blob/master/endpoints/openrtb2/amp_auction.go#L132

While both objects are similar at the moment, they provide a separation of concerns that gives us the ability to evolve each of them separately thus, keeping each of the components abstracted from specifics of each layer:

-  _EventRequest_ is the object within the endpoint's scope and from where we extract information to build our NotificationEvent (we could enrich our NotificationEvent with data from other sources besides _EventRequest_).

-  _NotificationEvent_ is the object from the Analytics domain and the one that should to be passed to the event handlers.

Happy to hear your thoughts on this.